File: tools/tailor_cv_ts/src/llm_gemini.ts
import { GoogleGenAI } from "@google/genai";
import { logger } from "./logger";

export const DEFAULT_GEMINI_MODEL = "gemini-2.0-flash";

export async function callGemini(
  system: string,
  userMessage: string,
  model = DEFAULT_GEMINI_MODEL
): Promise<string> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey)
    throw new Error("GEMINI_API_KEY environment variable is not set");

  logger.info(`Calling Gemini model '${model}'`);
  logger.debug(`System:\n${system}`);
  logger.debug(`User message:\n${userMessage}`);

  const ai = new GoogleGenAI({ apiKey });
  const response = await ai.models.generateContent({
    model,
    contents: userMessage,
    config: {
      systemInstruction: system,
      temperature: 0.2,
      topK: 40, // Allows for varied vocabulary (good for action verbs)
      topP: 0.85, // Prevents robotic repetition
      maxOutputTokens: 8192, // Safely accommodates a full 2-page Markdown CV
    },
  });

  const result = response.text ?? "";
  logger.debug(`Response:\n${result}`);
  return result;
}

File: tools/tailor_cv_py/src/tailor_cv/llm_gemini.py
import os

from google import genai
from google.genai import types
from loguru import logger

DEFAULT_GEMINI_MODEL = "gemini-2.0-flash"

GEMINI_OPTIONS = types.GenerateContentConfig(
    temperature=0.0,
    max_output_tokens=2048,
)


def call_gemini(system: str, user_message: str, model: str = DEFAULT_GEMINI_MODEL) -> str:
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        raise RuntimeError("GEMINI_API_KEY environment variable is not set")

    logger.debug(f"System:\n{system}")
    logger.debug(f"User message:\n{user_message}")
    logger.info(f"Calling Gemini model '{model}'")

    client = genai.Client(api_key=api_key)
    response = client.models.generate_content(
        model=model,
        config=types.GenerateContentConfig(
            system_instruction=system,
            temperature=GEMINI_OPTIONS.temperature,
            max_output_tokens=GEMINI_OPTIONS.max_output_tokens,
        ),
        contents=user_message,
    )

    result = response.text
    logger.debug(f"Response:\n{result}")
    return result
